Exec/hydro_tests/gamma_law_bubble/Problem_Derive.cpp

```diff
@@ -1,99 +1,390 @@
+
+  });
+
+}
+
+
+void ca_derrhopert(const Box& bx, FArrayBox& derfab, int dcomp, int /*ncomp*/,
+                    const FArrayBox& datfab, const Geometry& geomdata,
+                    Real /*time*/, const int* /*bcrec*/, int /*level*/)
+
+{
+
+  const auto dx = geomdata.CellSizeArray();
+  const auto problo = geomdata.ProbLoArray();
+
+  GpuArray<Real, 3> center;
+  ca_get_center(center.begin());
+
+  auto const dat = datfab.array();
+  auto const der = derfab.array();
+
+  // first make a 1D initial model for the entire domain
+  int npts_1d = (2.0_rt*center[AMREX_SPACEDIM-1] + 1.e-8_rt) /
+    dx[AMREX_SPACEDIM-1];
+
+  RealVector density_rv(npts_1d, 0);
+  Real* const density = density_rv.dataPtr();
+
+  density[0]  = dens_base;
+
+  // only initialize the first species
+  Real xn[NumSpec];
+  xn[0] = 1.0_rt;
+
+  // compute the pressure scale height (for an isothermal, ideal-gas
+  // atmosphere)
+  Real H = pres_base / dens_base / std::abs(gravity::const_grav);
+
+  for (int j = 0; j < npts_1d; j++) {
+
+     if (do_isentropic) {
+       Real z = static_cast<Real>(j) * dx[AMREX_SPACEDIM-1];
+       density[j] = dens_base * 
+         std::pow((gravity::const_grav * dens_base * (eos_gamma - 1.0_rt) * z/
+                   (eos_gamma*pres_base) + 1.0_rt), 1.0_rt/(eos_gamma - 1.0_rt));
+     } else {
+       Real z = (static_cast<Real>(j) + 0.5_rt) * dx[AMREX_SPACEDIM-1];
+       density[j] = dens_base * std::exp(-z/H);
+     }
+
+  }
+
+  amrex::ParallelFor(bx,
+  [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+  {
+    der(i,j,k,0) = dat(i,j,k,URHO) - density[AMREX_D_PICK(i,j,k)];
+  });
+
+}
+
+
+void ca_dertpert(const Box& bx, FArrayBox& derfab, int dcomp, int /*ncomp*/,
+                 const FArrayBox& datfab, const Geometry& geomdata,
+                 Real /*time*/, const int* /*bcrec*/, int /*level*/)
+{
+
+  const auto dx = geomdata.CellSizeArray();
+  const auto problo = geomdata.ProbLoArray();
+
+  GpuArray<Real, 3> center;
+  ca_get_center(center.begin());
+
+  auto const dat = datfab.array();
+  auto const der = derfab.array();
+
+  // first make a 1D initial model for the entire domain
+  int npts_1d = (2.0_rt*center[AMREX_SPACEDIM-1] + 1.e-8_rt) /
+    dx[AMREX_SPACEDIM-1];
+
+  RealVector pressure_rv(npts_1d, 0);
+  RealVector density_rv(npts_1d, 0);
+  RealVector temp_rv(npts_1d, 0);
+  RealVector eint_rv(npts_1d, 0);
+
+  Real* const pressure = pressure_rv.dataPtr();
+  Real* const density = density_rv.dataPtr();
+  Real* const temp = temp_rv.dataPtr();
+  Real* const eint = eint_rv.dataPtr();
+
+  pressure[0] = pres_base;
+  density[0]  = dens_base;
+
+  // only initialize the first species
+  Real xn[NumSpec];
+  xn[0] = 1.0_rt;
+
+  // compute the pressure scale height (for an isothermal, ideal-gas
+  // atmosphere)
+  Real H = pres_base / dens_base / std::abs(gravity::const_grav);
+
+  for (int j = 0; j < npts_1d; j++) {
```

Comment:
This looks fine but we repeat this multiple times so I wonder if we could later create a function to do this.

Reply:
I put this code in a inline function in a new `prob_util.H`